File: DSA/Hima-Swarupa-Padala/day20q2.py
import heapq

def dijkstra(graph, start):

    distances = [float('inf')] * len(graph)
    distances[start] = 0

    min_heap = [(0, start)]
    
    while min_heap:

        curr_dist, curr_node = heapq.heappop(min_heap)

        for neighbor, weight in graph[curr_node]:
            new_dist = curr_dist + weight
            if new_dist < distances[neighbor]:
                distances[neighbor] = new_dist
                heapq.heappush(min_heap, (new_dist, neighbor))
    
    return distances

def main():

    vertices = 4
    edges = [(1, 2, 4), (1, 3, 1), (2, 4, 2), (3, 4, 3)]
    start_node = 1
    
    graph = [[] for _ in range(vertices)]
    for u, v, w in edges:
        graph[u - 1].append((v - 1, w))
        graph[v - 1].append((u - 1, w))
    
    shortest_distances = dijkstra(graph, start_node - 1)

    print(shortest_distances)

if __name__ == "__main__":
    main()

File: DSA/Ishita-Nigam/Day20/Day20_Q2.java
import java.util.*;

class Day20_Q2 {
    static class Edge {
        int target, weight;

        Edge(int target, int weight) {
            this.target = target;
            this.weight = weight;
        }
    }

    public static int[] dijkstra(int vertices, List<Edge>[] graph, int start) {
        int[] distances = new int[vertices];
        Arrays.fill(distances, Integer.MAX_VALUE);
        distances[start] = 0;

        PriorityQueue<Edge> pq = new PriorityQueue<>(Comparator.comparingInt(e -> e.weight));
        pq.add(new Edge(start, 0));

        while (!pq.isEmpty()) {
            Edge current = pq.poll();
            int currentNode = current.target;

            for (Edge edge : graph[currentNode]) {
                int newDist = distances[currentNode] + edge.weight;
                if (newDist < distances[edge.target]) {
                    distances[edge.target] = newDist;
                    pq.add(new Edge(edge.target, newDist));
                }
            }
        }

        return distances;
    }

    public static void main(String[] args) {
        int vertices = 4;
        int edgesCount = 4;
        int startNode = 0;

        List<Edge>[] graph = new List[vertices];
        for (int i = 0; i < vertices; i++) {
            graph[i] = new ArrayList<>();
        }

        graph[0].add(new Edge(1, 4));
        graph[0].add(new Edge(2, 1));
        graph[1].add(new Edge(3, 2));
        graph[2].add(new Edge(3, 3));

        int[] distances = dijkstra(vertices, graph, startNode);

        System.out.println("Shortest path distances from node " + (startNode + 1) + ":");
        for (int i = 0; i < distances.length; i++) {
            System.out.println("To node " + (i + 1) + ": " + distances[i]);
        }
    }
}
